connect-inject/endpoints_controller.go

```diff
@@ -0,0 +1,429 @@
+		}
+
+		// Deregister each service instance that matches the metadata.
+		for svcID, serviceRegistration := range svcs {
+			// If we selectively deregister, only deregister if the address is not in the map. Otherwise, deregister
+			// every service instance.
+			if selectivelyDeregister {
+				if _, ok := endpointsAddressesMap[serviceRegistration.Address]; !ok {
+					// If the service address is not in the Endpoints addresses, deregister it.
+					if err = client.Agent().ServiceDeregister(svcID); err != nil {
+						r.Log.Error(err, "failed to deregister service instance", "consul-service-id", svcID)
+						return err
+					}
+				}
+			} else {
+				if err = client.Agent().ServiceDeregister(svcID); err != nil {
+					r.Log.Error(err, "failed to deregister service instance", "consul-service-id", svcID)
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// processUpstreams reads the list of upstreams from the Pod annotation and converts them into a list of api.Upstream
+// objects.
+func (r *EndpointsController) processUpstreams(pod *corev1.Pod) ([]api.Upstream, error) {
+	var upstreams []api.Upstream
+	if raw, ok := pod.Annotations[annotationUpstreams]; ok && raw != "" {
+		for _, raw := range strings.Split(raw, ",") {
+			parts := strings.SplitN(raw, ":", 3)
+
+			var datacenter, serviceName, preparedQuery string
+			var port int32
+			if strings.TrimSpace(parts[0]) == "prepared_query" {
+				port, _ = portValue(pod, strings.TrimSpace(parts[2]))
+				preparedQuery = strings.TrimSpace(parts[1])
+			} else {
+				port, _ = portValue(pod, strings.TrimSpace(parts[1]))
+
+				// TODO: Parse the namespace if provided
+
+				serviceName = strings.TrimSpace(parts[0])
+
+				// parse the optional datacenter
+				if len(parts) > 2 {
+					datacenter = strings.TrimSpace(parts[2])
+
+					// Check if there's a proxy defaults config with mesh gateway
+					// mode set to local or remote. This helps users from
+					// accidentally forgetting to set a mesh gateway mode
+					// and then being confused as to why their traffic isn't
+					// routing.
+					entry, _, err := r.ConsulClient.ConfigEntries().Get(api.ProxyDefaults, api.ProxyConfigGlobal, nil)
+					if err != nil && strings.Contains(err.Error(), "Unexpected response code: 404") {
+						return []api.Upstream{}, fmt.Errorf("upstream %q is invalid: there is no ProxyDefaults config to set mesh gateway mode", raw)
+					} else if err == nil {
+						mode := entry.(*api.ProxyConfigEntry).MeshGateway.Mode
+						if mode != api.MeshGatewayModeLocal && mode != api.MeshGatewayModeRemote {
+							return []api.Upstream{}, fmt.Errorf("upstream %q is invalid: ProxyDefaults mesh gateway mode is neither %q nor %q", raw, api.MeshGatewayModeLocal, api.MeshGatewayModeRemote)
+						}
+					}
+					// NOTE: If we can't reach Consul we don't error out because
+					// that would fail the pod scheduling and this is a nice-to-have
+					// check, not something that should block during a Consul hiccup.
```

Comment:
You're right, it would have prevented pod startup when it was in container_init as part of old service registration. However it won't now. Is there some other way we can help people realize they need to set a mesh gateway mode? Maybe we can just log rather than return an error here, to help people realize they need to set a mesh gateway mode?